src/reducers/url-state.js

```diff
@@ -481,6 +481,25 @@ const localTrackOrderByPid: Reducer<Map<Pid, TrackIndex[]>> = (
   }
 };
 
+const localTrackOrderChangedByPid: Reducer<Set<Pid>> = (
+  state = new Set(),
```

Comment:
Before looking at this patch, I was expecting to see a change to the code which computes the default order of threads if a profile is loaded that doesn't have an order in the URL. In that code, I was expecting to see a change which says "by default, no pids have a changed order".

But now I see that the only place with this default initialization is in the reducer. We're not actually touching the code that I was thinking of.

This might be fine but I wonder if we're missing any cases where we load a new profile and fail to reset the redux state. For example, what happens if I have a profile with a non-default order for a certain pid, and then I drag a new file into the profiler? What resets this state?

Reply:
> Before looking at this patch, I was expecting to see a change to the code which computes the default order of threads if a profile is loaded that doesn't have an order in the URL. In that code, I was expecting to see a change which says "by default, no pids have a changed order".
> 
> But now I see that the only place with this default initialization is in the reducer. We're not actually touching the code that I was thinking of.
> 
> This might be fine but I wonder if we're missing any cases where we load a new profile and fail to reset the redux state. For example, what happens if I have a profile with a non-default order for a certain pid, and then I drag a new file into the profiler? What resets this state?

That's a very good question!
In the PR text I wrote ` The good thing is that the unserializing logic was already working per-pid, so I didn't need to change it` but this wasn't well explained.

Let me try to explain this more.

1. what happens at load time:

This happens in https://github.com/firefox-devtools/profiler/blob/1d4b8a676b4086dacf15b0d805697490f9880971/src/profile-logic/tracks.js#L150-L180

Especially this is called in https://github.com/firefox-devtools/profiler/blob/1d4b8a676b4086dacf15b0d805697490f9880971/src/actions/receive-profile.js#L291 in the `finalizeFullProfileView` function, that's itself called from `loadProfile`.

In that code, we iterate over all pids, and if there's no track order defined for this pid in the URL, we use the default order.

So essentially that code was already working like we need for this patch!

2. What happens when I drag a new file?

Good question again!
These `localTrackOrder` (and others) properties are part of the "profile specific url state", they're all gathered in a subreducer object that's easy to reset at once. That's being reset in some cases in https://github.com/firefox-devtools/profiler/blob/1d4b8a676b4086dacf15b0d805697490f9880971/src/reducers/url-state.js#L662-L669.

Then the `loadProfile` is called again and it's been reinitialized again as said above.

I haven't actually tried this but I'll try it today. I'm confident this will work.

Reply:
I tried these 2 cases you're mentioning and they seem to work fine (ie, the localTrackOrderByPid is properly reset)

Reply:
Great, thank you for the explanation!